src/envoy/token/token_subscriber_test.cc

```diff
@@ -332,6 +332,58 @@ TEST_F(TokenSubscriberTest, Success) {
   ASSERT_TRUE(init_ready_);
 }
 
+TEST_F(TokenSubscriberTest, RetryMissingPreconditionThenSuccess) {
+  // Part 1: Failed due to missing precondition
+
+  // Setup mocks for info. Fail on the first time, then work later.
+  Envoy::Http::HeaderMapImplPtr req_headers{
+      new Envoy::Http::TestHeaderMapImpl{}};
+  EXPECT_CALL(*info_, prepareRequest(_))
+      .WillOnce(Return(ByMove(nullptr)))
+      .WillRepeatedly(
+          Return(ByMove(std::make_unique<Envoy::Http::RequestMessageImpl>(
+              std::move(req_headers)))));
+
+  // Setup fake parse status.
+  EXPECT_CALL(*info_, parseAccessToken(_, _))
+      .WillOnce(Invoke([](absl::string_view, TokenResult* ret) {
+        ret->token = "fake-token";
+        ret->expiry_duration = std::chrono::seconds(30);
+        return true;
+      }));
+
+  // Expect subscriber does not succeed at first, but then does.
+  EXPECT_CALL(*mock_timer_, enableTimer(kFailedExpect, nullptr)).Times(1);
+  EXPECT_CALL(*mock_timer_,
+              enableTimer(std::chrono::milliseconds(25 * 1000), nullptr))
+      .Times(1);
+  EXPECT_CALL(token_callback_, Call("fake-token")).Times(1);
+
+  // Start class under test.
+  setUp(TokenType::AccessToken);
+
+  // Assert subscriber did not succeed.
+  ASSERT_EQ(call_count_, 0);
+  ASSERT_FALSE(init_ready_);
+
+  // Part 2: Retry, will result in a success.
+  init_target_handle_->initialize(init_watcher_);
```

Comment:
My understanding is:  this init call starts everything.   we don't do anything before this call. so expecting call_count as 0 is not testing anything.

we should use client_callback->onFailure or On Success to control the flow.

I am not sure if the whole test is doing this.  Could you double check it?

Reply:
Line 363 `setUp` will actually start the class by calling `initialize()`. Then the async mock will kick in, the rest of the class with be executed.

Is there a way to make this more clear in our test case?

Reply:
In Envoy initalization flow, Envoy never call init_target_handle_->initialize() twice.  It is already called in setup(), so this call is not the same as production flow.

Reply:
Discussed offline. Captured the callback for the timer, done.